File: 1922-count-good-numbers/1922-count-good-numbers.java
class Solution {
    int mod = 1000000007;

    public int countGoodNumbers(long n) {
        return (int)(solve(5, (n + 1) / 2L)%mod * solve(4, (n / 2L))%mod)%mod;
    }

  long solve(int x, long n) {
        if (n == 0) {
                return 1;
        }
        long half=solve(x,n/2)%mod;
        if(n%2==1){
            return (x*((half*half)%mod))%mod;
        }
        return (half*half)%mod;
    }
}

File: 1922-count-good-numbers/1922-count-good-numbers.py
class Solution:
    def __init__(self):
        self.mod = 1000000007

    def countGoodNumbers(self, n: int) -> int:
        return (self.solve(5, (n + 1) // 2) % self.mod * 
                self.solve(4, n // 2) % self.mod) % self.mod

    def solve(self, x: int, n: int) -> int:
        if n == 0:
            return 1
        
        half = self.solve(x, n // 2) % self.mod
        
        if n % 2 == 1:
            return (x * (half * half % self.mod)) % self.mod
        
        return (half * half) % self.mod